oci/default.nix

```diff
@@ -0,0 +1,558 @@
+      # structure
+      if [ -f $out/layer.tar ]; then
+        ${python3}/bin/python ${./hash-layer.py} $out
+      fi
+    '';
+
+  /* Creates an OCI layer
+
+     entries is an attrset describing files to create / copy into the layer.
+     The attrset name will be the path of the entry, and the value is an
+     attrset describing it.
+
+     The entry attrset MUST have a type which is one of "file", "directory" or
+     "link". "mode", "uid", "gid" may also be specified to change the file
+     attributes where applicable.
+
+     File entries may have either a text attr which writes the contents as is
+     to the file, or a source attr which specifies a path to copy from.
+     Directory entries can have a sources attr which is a list of paths to copy
+     from. Each source is an attrset containing a path and uid, gid, and mode
+     specifier (0755, +w, etc). Link entries must have a target attr specifying
+     where to link to.
+
+     umask specifies the "umask" to be applied to entries by default.
+
+     command is a string that lets you create a layer from the contents of the
+     current directory after running the command. Note that you always start
+     with an empty directory, this does not let you modify existing layers.
+     The command also runs in the builder so restrictions such as no chown-ing
+     apply.
+
+     path is a list of derivations to add to the PATH of the resulting image
+     containing this layer.
+
+     includes is a list of derivations to include into the layer's nix store.
+     This is mostly useful for putting less frequently changing dependencies
+     earlier in the layer list.
+
+     excludes is a list of derivations whose closure should be excluded from
+     the layer (in the case where those derivations are already included in an
+     earlier layer).
+  */
+  makeLayer = {
+    name,
+    entries ? null,
+    umask ? "0222",
+    path ? [],
+    includes ? [],
+    excludes ? [],
+  }:
+    let
+      # we build the bareLayer first so that we can include it in the closure
+      # calculation, this ensures that any referenced store paths in entries is
+      # included in the full layer
+      #
+      # this works because the tar isn't compressed and the references are
+      # plainly visible
+      bareLayer = lib.optionals (entries != null && entries != {}) [(
+        makeBareLayerFromEntries {
+          name = "${name}-bare";
+          inherit entries umask;
+        }
+      )];
+
+      fullIncludes = path ++ includes ++ bareLayer;
+    in
+    makeLayer' {
+      inherit name excludes;
+
+      baseTar = map (layer: "${layer}/layer.tar") bareLayer;
+
+      includes = fullIncludes;
+
+      passthru = {
+        inherit path;
+        bare = bareLayer;
+      };
+    };
+
+  /* Creates an OCI image manifest and config
+
+     architecture and os should be one of GOARCH and GOOS as listed in
+     https://golang.org/doc/install/source#environment
```

Comment:
I've not tested this with cross-arch builds. I'm not sure how to begin with this, do you have any suggestions?